uarray/_uarray_dispatch.cxx

```diff
@@ -822,15 +822,96 @@ int Function_traverse(Function * self, visitproc visit, void * arg)
 }
 
 
+/** Break reference cycles when being GCed */
+int Function_clear(Function * self)
+{
+  self->extractor_.reset();
+  self->replacer_.reset();
+  self->def_args_.reset();
+  self->def_kwargs_.reset();
+  self->def_impl_.reset();
+  self->dict_.reset();
+  return 0;
+}
+
+
+/** Support for pickle.dump */
+PyObject * Function___getstate__(Function * self, PyObject * /*args*/)
+{
+  auto domain = py_ref::steal(
+    PyUnicode_FromStringAndSize(self->domain_key_.data(),
+                                self->domain_key_.size()));
+  if (!domain)
+    return nullptr;
+
+  auto res = py_make_tuple(self->extractor_,
+                           self->replacer_,
+                           domain,
+                           self->def_args_,
+                           self->def_kwargs_,
+                           self->def_impl_,
+                           self->dict_ ? self->dict_ : Py_None);
+
+  return res.release();
+}
+
+
+/** Support for pickle.load */
+PyObject * Function___setstate__(Function * self, PyObject * args)
+{
+  PyObject * state_tuple = nullptr;
+  if (!PyArg_ParseTuple(args, "O!", &PyTuple_Type, &state_tuple))
+    return nullptr;
+
+  if (PyTuple_GET_SIZE(state_tuple) != 7)
+  {
+    PyErr_SetString(PyExc_ValueError, "State tuple must have 7 entries");
+    return nullptr;
+  }
+
+  auto init_args = py_ref::steal(PyTuple_GetSlice(state_tuple, 0, 6));
+  if (!init_args)
+    return nullptr;
+
+  if (Function::init(self, init_args, nullptr) < 0)
+    return nullptr;
+
+  auto dict = py_ref::ref(PyTuple_GET_ITEM(state_tuple, 6));
+  if (dict == Py_None)
+  {
+    self->dict_.reset();
+  }
+  else
+  {
+    if (!PyDict_CheckExact(dict))
+    {
+      PyErr_SetString(PyExc_TypeError, "State __dict__ is not a dict object");
+      return nullptr;
+    }
+
+    self->dict_ = std::move(dict);
+  }
+
+  Py_RETURN_NONE;
+}
+
+
 PyGetSetDef Function_getset[] =
 {
   {"__dict__", PyObject_GenericGetDict, PyObject_GenericSetDict},
   {NULL} /* Sentinel */
 };
 
+PyMethodDef Function_methods[] =
+{
+  {"__getstate__", (binaryfunc)Function___getstate__, METH_NOARGS, nullptr},
+  {"__setstate__", (binaryfunc)Function___setstate__, METH_VARARGS, nullptr},
+  {NULL} /* Sentinel */
+};
+
 PyTypeObject FunctionType = {
   PyVarObject_HEAD_INIT(NULL, 0)
-  "_uarray.Function",             /* tp_name */
+  "uarray._uarray.Function",      /* tp_name */
```

Comment:
I feel that the name of a private submodule shouldn't be exposed, NumPy does the same. Is it possible to change all these to `uarray.*` instead of `uarray._uarray.*`?

Reply:
Do you want `uarray._Function` or do you think the type should be public?

Reply:
Private. The default args/kwargs are passed into `__init__` so that could make things complicated for the user and misuse easy.